[LoopVectorize] Enable hoisting of runtime checks by default

david-arm opened the pull request:
With commit https://reviews.llvm.org/D152366 I introduced functionality that permitted the hoisting of runtime memory checks from a vectorised inner loop to the preheader of the next outer-most loop. This is useful for benchmarks like SPEC2017's x264 where the inner loop is vectorised and only has a small trip count. In such cases the runtime memory checks become expensive and since the checks never fail in the case of x264 it makes sense to do this. However, this behaviour was controlled by the flag -hoist-runtime-checks which was off by default.

This patch enables this flag by default for all targets, since I believe this is a generally beneficial thing to do. I have tested this with SPEC2017 and I see 2.3% and 2.6% improvements with x264 on neoverse-v1 and neoverse-n1, respectively. Similarly, I saw slight improvements in the overall geomean on both machines. The only other notable changes were a 1% drop in the roms benchmark, which was compensated for by a 1% improvement in fotonik3d.

david-arm committed:
[LoopVectorize] Enable hoisting of runtime checks by default

With commit https://reviews.llvm.org/D152366 I introduced
functionality that permitted the hoisting of runtime memory checks
from a vectorised inner loop to the preheader of the next outer-most
loop. This is useful for benchmarks like SPEC2017's x264 where the
inner loop is vectorised and only has a small trip count. In such
cases the runtime memory checks become expensive and since the checks
never fail in the case of x264 it makes sense to do this. However,
this behaviour was controlled by the flag -hoist-runtime-checks
which was off by default.

This patch enables this flag by default for all targets, since I
believe this is a generally beneficial thing to do. I have tested
this with SPEC2017 and I see 2.3% and 2.6% improvements with x264 on
neoverse-v1 and neoverse-n1, respectively. Similarly, I saw slight
improvements in the overall geomean on both machines. The only
other notable changes were a 1% drop in the roms benchmark, which
was compensated for by a 1% improvement in fotonik3d.

davemgreen commented on llvm/test/Transforms/LoopVectorize/invariant-store-vectorization.ll:
```diff
@@ -13,9 +13,6 @@ target datalayout = "e-p:64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f3
 ; address.
 
 
-; memory check is found.conflict = b[max(n-1,1)] > a && (ptr a)+1 > (ptr b)
```
Why is this removed? It looks like there is a memcheck below.

david-arm replied:
That's a good question! I wasn't sure what to do here, because the exact contents of found.conflict depend upon whether you're trying to hoist the runtime memory checks or not. Perhaps it was just lazy of me to remove it - I'll update the comment to reflect what's it actually now checking.

davemgreen reviewed (commented):
I can confirm the embedded benchmarks I have tried don't show any problems from this after the recent fix, they were all improving where they changed.

david-arm committed:
Re-add comment in invariant-store-vectorization.ll

In the original version of my patch I removed a comment about
memory checks, thinking that the check had now changed with
my patch. However, the comment refers to a function that does
not have a nested loop, so the checks don't actually change.

fhahn commented:
Thanks for sharing the patch! I'll run some perf tests on my side, should hopefully be done by Monday.

david-arm commented:
> Thanks for sharing the patch! I'll run some perf tests on my side, should hopefully be done by Monday.

Hi @fhahn, good morning! Did you get chance to look at this?

fhahn commented:
> > Thanks for sharing the patch! I'll run some perf tests on my side, should hopefully be done by Monday.
> 
> Hi @fhahn, good morning! Did you get chance to look at this?

Hi, sorry for the delay, there was some infrastructure issues with our perf testing infrastructure. But the benchmarks are running now, will share the results soon hopefully. Please ping me again tomorrow if I forget the respond by then :)

fhahn commented:
There's one internal benchmark where there may be a ~2% regression with this patch (and extra vectorizer passes enabled).  I'll confirm tomorrow if this is actually caused by the patch or noise only

david-arm commented:
> I had a look and there are some cases where the more lightweight diff-checks are more profitable than the hoisted checks after [32d1197](https://github.com/llvm/llvm-project/commit/32d1197a8faf4c04df5fcd6c0588aa288bc42e35).
> 
> One particular case is when the start of the AddRecs in the inner loop is the same AddRec in the outer loop, like in
> 
> https://github.com/llvm/llvm-project/blob/fd9a777e018d23d258646af613179eebf6df4b34/llvm/test/Transforms/LoopVectorize/runtime-checks-difference.ll#L279
> 
> It would probably be good to refine the checks when to pick hoisting vs diff checks

Hmm, I guess the trick is to do this in a way that doesn't make x264 slower because I know one of the loops in x264 that benefits from this do have lightweight diff-checks. Even diff checks are still expensive for short inner loops. Do you know of any specific benchmarks where this is a problem? Is https://github.com/llvm/llvm-project/blob/fd9a777e018d23d258646af613179eebf6df4b34/llvm/test/Transforms/LoopVectorize/runtime-checks-difference.ll#L279 the most definitive example of benchmarks you're worried will get slower?

Anyway, I'm happy to take a further look at this and see if I can get the best of both worlds. If necessary the only alternative is to enable this under a TTI hook, but I was hoping to avoid that because it's a very invasive change.

fhahn commented:
> Hmm, I guess the trick is to do this in a way that doesn't make x264 slower because I know one of the loops in x264 that benefits from this do have lightweight diff-checks. Even diff checks are still expensive for short inner loops. Do you know of any specific benchmarks where this is a problem? Is
> 
> https://github.com/llvm/llvm-project/blob/fd9a777e018d23d258646af613179eebf6df4b34/llvm/test/Transforms/LoopVectorize/runtime-checks-difference.ll#L279
> 
> the most definitive example of benchmarks you're worried will get slower?

For cases like this one (where the start of the inner AddRecs is the same outer loop AddRec), the difference should also become loop invariant, as the diff checks subtract the starts, so if the outer AddRecs are the same this simplifies to the start of the outer AddRec, which is invariant.

david-arm commented:
Ah, thanks @fhahn that makes sense then!

david-arm commented:
Hi @fhahn, I've created https://github.com/llvm/llvm-project/pull/73515 to prefer diff checks if we can prove the diff check will itself be invariant in the outer loop.

david-arm committed:
Merge branch 'llvm:main' into hoist_rc

david-arm commented:
Gentle ping! https://github.com/llvm/llvm-project/pull/73515 has now landed so I think this patch should be ready to go.

davemgreen reviewed (approved):
With that fixed, and from the perf Ive seen, this LGTM. Thanks

fhahn reviewed (approved):
LGTM with the earlier adjustments, thanks!
